### DISCUSSION TITLE
Is there getter/setter generation?

### DISCUSSION THREAD

**Discussion by couryrr** [mfussenegger/nvim-jdtls#504] [Q&A]
I have looked through the docs and do not see anything about this feature specifically.

**Comment by gbroques** [comment]
> I have looked through the docs and do not see anything about this feature specifically.

Generate getters / setters are available as code actions.

Maybe we could add those as sub-bullets to the "Code action extensions" bullet in the README?

https://github.com/mfussenegger/nvim-jdtls/blob/b041dd4308d8c715187274ece9bc8d501a6f3041/README.md?plain=1#L28-L35

**Reply by mfussenegger** [reply]
> Maybe we could add those as sub-bullets to the "Code action extensions" bullet in the README?

The readme only lists extensions that nvim-jdtls adds. Getter & setter support works out of the box with the stock neovim core lsp client and doesn't depend on an additional client extension.

**Reply by couryrr** [reply]
I might be dense but I am not seeing this. Do you have a link to where this configuration is?

**Reply by gbroques** [reply]
> I might be dense but I am not seeing this. Do you have a link to where this configuration is?

@couryrr Not seeing what?

There's a ["Code action extensions" section in the README](https://github.com/mfussenegger/nvim-jdtls/blob/b041dd4308d8c715187274ece9bc8d501a6f3041/README.md?plain=1#L28-L35), but as Mathias said, it only documents **extensions** to code actions that `nvim-jdtls` has code for.

Getters / setters generation work as code actions "out of the box" for `nvim-jdtls` — so they aren't documented in this project.

Code actions from Neovim's builtin LSP client can be triggered via `vim.lsp.buf.code_action`.

You'll want to map this to something, for example `<leader>la` as a mnemonic for "**L**SP **a**ctions":
https://github.com/gbroques/neovim-configuration/blob/a2aae82211d0a5fe5f22fd4e2c1d21310ffec755/lua/keymaps.lua#L89

`vim.lsp.buf.code_action` is documented by Neovim here:
https://neovim.io/doc/user/lsp.html#vim.lsp.buf.code_action()

---

As an aside, there's a brief mention of "**advanced** getters / setters generation" in the "Extended Client Capabilities" section of the wiki for JDTLS:
https://github.com/eclipse/eclipse.jdt.ls/wiki/Language-Server-Settings-&-Capabilities#extended-client-capabilities

> `advancedGenerateAccessorsSupport` - Client supports using a prompt to allow a user to configure behaviour for generating Getters/Setters.

I've setup `nvim-jdtls` to use these extended client capabilities:
https://github.com/gbroques/neovim-configuration/blob/a2aae82211d0a5fe5f22fd4e2c1d21310ffec755/ftplugin/java.lua#L3

I'm not sure if the extended client capabilities are needed for "basic" getters / setters generation, but it's what I have configured.

---

Hopefully that helps.

**Reply by couryrr** [reply]
I am not seeing anything about generating getters and setters. I have `vim.lsp.buf.code_action` mapped to `<leader>ca` and that action is not referenced when I page through.

**Reply by gbroques** [reply]
Are you in a class with a field to generate getters / setters for?

For example:
```java
class Foo {
    private String baz;
}
```

I would put your cursor on the field (e.g. `baz`) before executing `vim.lsp.buf.code_action`.

Like I mentioned earlier, I have "extended client capabilities" enabled.

Maybe try doing the same.

**Reply by couryrr** [reply]
Success, not exactly sure what I changed in all my toiling over the weekend but I get the context option when on a field now.
